tapdb/multiverse.go

```diff
@@ -61,21 +70,306 @@ type BatchedMultiverse interface {
+type cachedProof []*universe.Proof
+
+// Size just returns 1 as we're limiting based on the total number of proofs.
+func (c *cachedProof) Size() (uint64, error) {
+	return 1, nil
+}
+
+// leafProofCache is used to cache proofs for issuance leaves for assets w/o a
+// group key.
+type leafProofCache = *lru.Cache[ProofKey, *cachedProof]
+
+// newLeafCache creates a new leaf proof cache.
+func newLeafCache() leafProofCache {
+	return lru.NewCache[ProofKey, *cachedProof](
+		numCachedProofs,
+	)
+}
+
+// treeID is used to uniquely identify a multiverse tree.
+type treeID string
+
+// proofCache a map of proof caches for each proof type.
+type proofCache lnutils.SyncMap[treeID, leafProofCache]
+
+// newProofCache creates a new proof cache.
+func newProofCache() *proofCache {
+	return &proofCache{}
+}
+
+// fetchProof reads the cached proof for the given ID and leaf key.
+func (p *proofCache) fetchProof(id universe.Identifier,
+	leafKey universe.LeafKey) []*universe.Proof {
+
+	// First, get the sub-cache for this universe ID from the map of
+	// caches.
+	idStr := treeID(id.String())
+	proofCache, _ := p.LoadOrStore(idStr, newLeafCache())
+	assetProofCache := proofCache.(leafProofCache)
+
+	// With that lower level cache obtained, we can check to see if we have
+	// a hit or not.
+	proofKey := NewProofKey(id, leafKey)
+	proof, err := assetProofCache.Get(proofKey)
+	if err == nil {
+		log.Debugf("read proof for %v+%v from cache, key=%v",
+			id.StringForLog(), leafKey, proofKey[:])
+		return *proof
+	}
+
+	return nil
+}
+
+// insertProof inserts the given proof into the cache.
+func (p *proofCache) insertProof(id universe.Identifier, leafKey universe.LeafKey,
+	proof []*universe.Proof) {
+
+	idStr := treeID(id.String())
+
+	proofCache, _ := p.LoadOrStore(idStr, newLeafCache())
+	assetProofCache := proofCache.(leafProofCache)
+
+	proofKey := NewProofKey(id, leafKey)
+
+	log.Debugf("storing proof for %v+%v in cache, key=%x",
+		id.StringForLog(), leafKey, proofKey[:])
+
+	proofVal := cachedProof(proof)
+	assetProofCache.Put(proofKey, &proofVal)
+}
+
+// delProofsForAsset deletes all the proofs for the given asset.
+func (p *proofCache) delProofsForAsset(id universe.Identifier) {
+	log.Debugf("wiping proofs for %v from cache", id)
+
+	idStr := treeID(id.String())
+	p.Delete(idStr)
+}
+
+// cachedRoots is used to store the latest root for each known universe tree.
+type cachedRoots []universe.Root
+
+// atomicRootCache is an atomic pointer to a root cache.
+type atomicRootCache = atomic.Pointer[cachedRoots]
+
+// newAtomicRootCache creates a new atomic root cache.
+func newAtomicRootCache() atomicRootCache {
+	treeCache := &cachedRoots{}
+
+	var a atomicRootCache
+	a.Store(treeCache)
+
+	return a
+}
+
+// rootIndex maps a tree ID to a universe root.
+type rootIndex = lnutils.SyncMap[treeID, *universe.Root]
+
+// atomicRootIndex is an atomic pointer to a root index.
+type atomicRootIndex = atomic.Pointer[rootIndex]
```

Comment:
I guess these indexes backed by `SyncMap`s don't necessarily need a size limit as they are getting wiped p often?

Also, here the `atomic.Pointer` is basically acting as a mutex + pointer? Vs. only needing a `Mutex` for the new `lru.Cache` objects.

Reply:
Yeah exactly, using the atomic pointer we can just hot swap the pointer value to nil/empty, which then triggers the cache repopulation.